src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                    return Add((DecisionTree.ByValue)decision, makeDecision);
+                case DecisionTree.DecisionKind.Guarded:
+                    return Add((DecisionTree.Guarded)decision, makeDecision);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(decision.Kind);
+            }
+        }
+
+        private DecisionTree Add(DecisionTree.Guarded guarded, DecisionMaker makeDecision)
+        {
+            if (guarded.Default != null)
+            {
+                if (guarded.Default.MatchIsComplete) return null;
+                var result = Add(guarded.Default, makeDecision);
+                if (guarded.Default.MatchIsComplete) guarded.MatchIsComplete = true;
+                return result;
+            }
+            else
+            {
+                var result = guarded.Default = makeDecision(guarded.Expression, guarded.Type);
+                guarded.Default.Parent = guarded;
+                if (guarded.Default.MatchIsComplete) guarded.MatchIsComplete = true;
+                return result;
+            }
+        }
+
+        private DecisionTree Add(DecisionTree.ByValue byValue, DecisionMaker makeDecision)
+        {
+            DecisionTree result;
+            if (byValue.Default != null)
+            {
+                result = Add(byValue.Default, makeDecision);
+            }
+            else
+            {
+                result = byValue.Default = makeDecision(byValue.Expression, byValue.Type);
+                result.Parent = byValue;
+            }
+            if (byValue.Default.MatchIsComplete) byValue.MatchIsComplete = true;
+            return result;
+        }
+
+        private DecisionTree Add(DecisionTree.ByType byType, DecisionMaker makeDecision)
+        {
+            try
+            {
+                if (byType.Default == null)
+                {
+                    byType.Default = makeDecision(byType.Expression, byType.Type);
+                    byType.Default.Parent = byType;
+                    return byType.Default;
+                }
+                else
+                {
+                    return Add(byType.Default, makeDecision);
+                }
+            }
+            finally
+            {
+                if (byType.Default.MatchIsComplete) byType.MatchIsComplete = true;
+            }
+        }
+    }
+
+    internal static class PatternConversionExtensions
+    {
+        // Does an expression of type expressionType "match" a pattern that looks for type patternType?
+        // 'true' if the matched type catches all of them, 'false' if it catches none of them, and
+        // 'null' if it might catch some of them. For this test we assume the expression's value
+        // isn't null.
+        public static bool? ExpressionOfTypeMatchesPatternType(this Conversions conversions, TypeSymbol expressionType, TypeSymbol patternType)
+        {
+            if (expressionType == patternType) return true;
```

Comment:
Yes. It is an optimization of the code below.

---
In reply to: [64983856](https://github.com/dotnet/roslyn/pull/11485#discussion_r64983856) [](ancestors = 64983856)